xds/src/main/java/io/grpc/xds/FaultFilter.java

```diff
@@ -0,0 +1,443 @@
+import io.envoyproxy.envoy.type.v3.FractionalPercent;
+import io.grpc.CallOptions;
+import io.grpc.Channel;
+import io.grpc.ClientCall;
+import io.grpc.ClientInterceptor;
+import io.grpc.Context;
+import io.grpc.Deadline;
+import io.grpc.LoadBalancer.PickSubchannelArgs;
+import io.grpc.Metadata;
+import io.grpc.MethodDescriptor;
+import io.grpc.Status;
+import io.grpc.internal.DelayedClientCall;
+import io.grpc.internal.GrpcUtil;
+import io.grpc.xds.FaultConfig.FaultAbort;
+import io.grpc.xds.FaultConfig.FaultDelay;
+import io.grpc.xds.Filter.ClientInterceptorBuilder;
+import io.grpc.xds.Matchers.HeaderMatcher;
+import io.grpc.xds.ThreadSafeRandom.ThreadSafeRandomImpl;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.concurrent.Executor;
+import java.util.concurrent.ScheduledExecutorService;
+import java.util.concurrent.ScheduledFuture;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.atomic.AtomicLong;
+import javax.annotation.Nullable;
+
+/** HttpFault filter implementation. */
+final class FaultFilter implements Filter, ClientInterceptorBuilder {
+
+  static final FaultFilter INSTANCE =
+      new FaultFilter(ThreadSafeRandomImpl.instance, new AtomicLong());
+  @VisibleForTesting
+  static final Metadata.Key<String> HEADER_DELAY_KEY =
+      Metadata.Key.of("x-envoy-fault-delay-request", Metadata.ASCII_STRING_MARSHALLER);
+  @VisibleForTesting
+  static final Metadata.Key<String> HEADER_DELAY_PERCENTAGE_KEY =
+      Metadata.Key.of("x-envoy-fault-delay-request-percentage", Metadata.ASCII_STRING_MARSHALLER);
+  @VisibleForTesting
+  static final Metadata.Key<String> HEADER_ABORT_HTTP_STATUS_KEY =
+      Metadata.Key.of("x-envoy-fault-abort-request", Metadata.ASCII_STRING_MARSHALLER);
+  @VisibleForTesting
+  static final Metadata.Key<String> HEADER_ABORT_GRPC_STATUS_KEY =
+      Metadata.Key.of("x-envoy-fault-abort-grpc-request", Metadata.ASCII_STRING_MARSHALLER);
+  @VisibleForTesting
+  static final Metadata.Key<String> HEADER_ABORT_PERCENTAGE_KEY =
+      Metadata.Key.of("x-envoy-fault-abort-request-percentage", Metadata.ASCII_STRING_MARSHALLER);
+  static final String TYPE_URL =
+      "type.googleapis.com/envoy.extensions.filters.http.fault.v3.HTTPFault";
+  private static final String TYPE_URL_TYPED_STRUCT =
+      "type.googleapis.com/udpa.type.v1.TypedStruct";
+
+  private final ThreadSafeRandom random;
+  private final AtomicLong activeFaultCounter;
+
+  @VisibleForTesting
+  FaultFilter(ThreadSafeRandom random, AtomicLong activeFaultCounter) {
+    this.random = random;
+    this.activeFaultCounter = activeFaultCounter;
+  }
+
+  @Override
+  public String[] typeUrls() {
+    return new String[] { TYPE_URL };
+  }
+
+  @Override
+  public StructOrError<FaultConfig> parseFilterConfig(Any rawProtoMessage) {
+    HTTPFault httpFaultProto;
+    try {
+      String typeUrl = rawProtoMessage.getTypeUrl();
+      if (typeUrl.equals(TYPE_URL_TYPED_STRUCT)) {
```

Comment:
Why this part needs to be implemented here? What's the precondition of this method? The `Any` argument should just be the type of "type.googleapis.com/envoy.extensions.filters.http.fault.v3.HTTPFault".

Reply:
If the rawProtoMessage's type url is `TYPE_URL_TYPED_STRUCT`, then it does not contain an `Any` message with type url "type.googleapis.com/envoy.extensions.filters.http.fault.v3.HTTPFault" inside it. It will contain a `TypedStruct` message, which basically is a json.

Reply:
The problem comes from the use of `String getFilterConfigTypeUrl(Any rawProtoMessage)` method. It _peeks_ the typeUrl by unpacking it and throwing away the unpacked result.

You should define the precondition of `parseFilterConfig(Any)` method such that the `Any` is of the type that defined by the `typeUrls()` method. You should not repeat the logic of peeking the typeUrl (once for checking if it is parsable by this Filter and once for actual parsing).

Reply:
Current API is identical to go https://github.com/grpc/grpc-go/blob/master/xds/internal/httpfilter/httpfilter.go#L43-L44

If you want to avoid repeating peek and unpack, the only way I can think of is to change the signature to 

```java
StructOrError<? extends FilterConfig> parseFilterConfig(Message message) {
  if (message instanceof Any) {
     ((Any) message).unpack(HTTPFault.class);
     ...
  } else if (message instanceof Struct) {
    JsonFormat.parser().merge(toJson((Struct) message), HTTPFault.newBuilder());
    ...
  } else {
    // throw error precondition
  }
}
```

Reply:
Remove this logic after being clarified that FaultFilter's config will not be typed_struct.